nav2_behavior_tree/plugins/action/remove_in_collision_goals_action.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2024 Angsa Robotics
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <limits>
+
+#include "nav2_behavior_tree/plugins/action/remove_in_collision_goals_action.hpp"
+#include "nav2_behavior_tree/bt_utils.hpp"
+#include "tf2/utils.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+
+namespace nav2_behavior_tree
+{
+
+RemoveInCollisionGoals::RemoveInCollisionGoals(
+  const std::string & service_node_name,
+  const BT::NodeConfiguration & conf)
+: BtServiceNode<nav2_msgs::srv::GetCosts>(service_node_name, conf),
+  use_footprint_(true),
+  cost_threshold_(253)
+{}
+
+
+void RemoveInCollisionGoals::on_tick()
+{
+  getInput("use_footprint", use_footprint_);
+  getInput("cost_threshold", cost_threshold_);
+
+  getInput("input_goals", input_goals_);
+
+  if (input_goals_.empty()) {
+    setOutput("output_goals", input_goals_);
+    should_send_request_ = false;
+    return;
+  }
+  request_ = std::make_shared<nav2_msgs::srv::GetCosts::Request>();
```

Comment:
The `request_` is never recreated/reset in bt_service_node. It would make sense to add that there don't you think?

Reply:
I think for this case since the goal vector length can change this is good!

Reply:
Sorry I wasn't clear enough, I meant to recreate the request in the BTServiceNode parent instead of for every child. Seems like a generic enough thing to do to me.

Reply:
Oh, I think some rely on it being the same by populating it once in initialization -- but if that's not the case, happy to have that change too. Some API change could come and make that break.

Reply:
It's only used [here](https://github.com/ros-navigation/navigation2/blob/746bb880a775201fbdc4cca38a873cc3d0d9ab40/nav2_behavior_tree/plugins/action/clear_costmap_service.cpp#L44) in Nav2 but some people might rely on it in their private plugins. I'm happy either way, the benefit of keeping it as is, is at least **to have the option** to initialize once.

Reply:
I'm ambivalent unless someone had a compelling argument either way.